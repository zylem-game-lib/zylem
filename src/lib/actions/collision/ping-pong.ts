import { CollisionContext, GameEntity } from "../../entities/entity";
import { MoveableEntity } from "../../behaviors/moveable";

export interface PingPongOptions {
	restitution: number;
	spinFactor: number;
	maxSpeed: number;
	minSpeed: number;
}

const defaultOptions: PingPongOptions = {
	restitution: 0.6,
	spinFactor: 0.5,
	maxSpeed: 15,
	minSpeed: 5
};

/**
 * Creates a realistic ricochet effect for ping-pong/ball physics
 * Takes into account the ball's current velocity and reflects it properly
 * 
 * @param options Configuration options for the ricochet physics
 * @param options.restitution Energy retention factor (0 = no bounce, 1 = perfect bounce)
 * @param options.spinFactor How much the paddle position affects the ricochet angle
 * @param options.maxSpeed Maximum velocity the ball can achieve
 * @param options.minSpeed Minimum velocity to maintain game momentum
 */
export function pingPong(
	options: Partial<PingPongOptions> = defaultOptions
) {
	return (collisionContext: CollisionContext<any, any>) => {
		_pingPong(collisionContext, options);
	};
}

function _pingPong(
	collisionContext: CollisionContext<MoveableEntity, GameEntity<any>>,
	options?: Partial<PingPongOptions>
) {
	const { entity, other } = collisionContext;
	if (other.collider?.isSensor()) return;

	const { restitution, spinFactor, maxSpeed, minSpeed } = { ...defaultOptions, ...options };

	const ballPos = entity.getPosition();
	const paddlePos = other.body?.translation();

	if (!ballPos || !paddlePos) return;

	const velocity = entity.getVelocity();
	if (!velocity) return;

	const normalX = ballPos.x > paddlePos.x ? 1 : -1;
	const normalY = 0;

	const paddleHeight = 2;
	const hitPosition = Math.max(-1, Math.min(1, (ballPos.y - paddlePos.y) / paddleHeight));

	const velocityDotNormal = velocity.x * normalX + velocity.y * normalY;
	let reflectedVelX = velocity.x - 2 * velocityDotNormal * normalX;
	let reflectedVelY = velocity.y - 2 * velocityDotNormal * normalY;

	reflectedVelY += hitPosition * spinFactor * Math.abs(reflectedVelX);

	reflectedVelX *= restitution;
	reflectedVelY *= restitution;

	const currentSpeed = Math.sqrt(reflectedVelX * reflectedVelX + reflectedVelY * reflectedVelY);
	if (currentSpeed < minSpeed) {
		const scale = minSpeed / currentSpeed;
		reflectedVelX *= scale;
		reflectedVelY *= scale;
	}

	if (currentSpeed > maxSpeed) {
		const scale = maxSpeed / currentSpeed;
		reflectedVelX *= scale;
		reflectedVelY *= scale;
	}

	const separation = 0.2;
	const newX = paddlePos.x + (normalX * separation);
	entity.setPosition(newX, ballPos.y, ballPos.z);

	entity.moveXY(reflectedVelX, reflectedVelY);
}
