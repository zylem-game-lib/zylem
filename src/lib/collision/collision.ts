import { RigidBodyDesc, RigidBodyType } from "@dimforge/rapier3d-compat";
import { Color } from "three";
import { Component, Entity, System } from "../core/ecs";

export class BaseCollision implements CollisionComponent {
	static: boolean = false;
	isSensor: boolean = false;
	bodyDescription: RigidBodyDesc;

	constructor({ isDynamicBody = true }) {
		const type = isDynamicBody ? RigidBodyType.Dynamic : RigidBodyType.Fixed;
		this.bodyDescription = new RigidBodyDesc(type)
			.setTranslation(0, 0, 0)
			.setGravityScale(1.0)
			.setCanSleep(false)
			.setCcdEnabled(true);
	}
}

export interface CollisionComponent extends Component {
	static: boolean;
	isSensor: boolean;
	bodyDescription: RigidBodyDesc;
};

export interface CollisionDebugComponent extends Component {
	active: boolean;
	color: Color;
}
