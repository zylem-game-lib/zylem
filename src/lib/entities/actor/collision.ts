import { ActiveCollisionTypes, ColliderDesc, RigidBodyDesc, RigidBodyType } from "@dimforge/rapier3d-compat";
import { BufferGeometry, Object3D, SkinnedMesh } from "three";
import { BaseCollision } from "~/lib/collision/_oldCollision";

export class ActorCollision extends BaseCollision {
	height: number = 1;
	radius: number = 1;

	createCollision({ isDynamicBody = true, object }: { isDynamicBody: boolean, object: Object3D | null }) {
		const type = isDynamicBody ? RigidBodyType.Dynamic : RigidBodyType.Fixed;
		this.bodyDescription = new RigidBodyDesc(type)
			.setTranslation(0, 0, 0)
			// .setRotation({ w: 1.0, x: 0.0, y: 0.0, z: 0.0 })
			.setGravityScale(1.0)
			.setCanSleep(false)
			.setCcdEnabled(false);
		if (!object) {
			console.warn('missing object');
			return;
		}
		const skinnedMesh = object.children[0] as SkinnedMesh;
		const geometry = skinnedMesh.geometry as BufferGeometry;

		if (!geometry) {
			return;
		}
		geometry.computeBoundingBox();
		if (geometry.boundingBox) {
			const maxY = geometry.boundingBox.max.y;
			const minY = geometry.boundingBox.min.y;
			this.height = maxY - minY;
		}
	}

	createCollider(isSensor: boolean = false) {
		// TODO: consider using offsets
		let colliderDesc = ColliderDesc.capsule(this.height / 2, 1);
		colliderDesc.setSensor(isSensor);
		colliderDesc.setTranslation(0, this.height / 2, 0);
		colliderDesc.activeCollisionTypes = (isSensor) ? ActiveCollisionTypes.KINEMATIC_FIXED : ActiveCollisionTypes.DEFAULT;
		return colliderDesc;
	}
}