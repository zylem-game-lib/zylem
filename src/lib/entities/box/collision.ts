import { ActiveCollisionTypes, ColliderDesc } from "@dimforge/rapier3d-compat";
import { Vector3 } from "three";
import { BaseCollision } from "~/lib/collision/_oldCollision";
import { SizeVector } from "~/lib/interfaces/utility";

export class BoxCollision extends BaseCollision {
	_size: SizeVector = new Vector3(1, 1, 1);

	createCollider(isSensor: boolean = false) {
		const size = this._size || new Vector3(1, 1, 1);
		const half = { x: size.x / 2, y: size.y / 2, z: size.z / 2 };
		let colliderDesc = ColliderDesc.cuboid(half.x, half.y, half.z);
		colliderDesc.setSensor(isSensor);
		colliderDesc.activeCollisionTypes = (isSensor) ? ActiveCollisionTypes.KINEMATIC_FIXED : ActiveCollisionTypes.DEFAULT;
		return colliderDesc;
	}
}