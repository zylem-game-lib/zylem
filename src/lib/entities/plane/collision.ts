import { ActiveCollisionTypes, ColliderDesc } from "@dimforge/rapier3d-compat";
import { BaseCollision } from "~/lib/collision/_oldCollision";

export class PlaneCollision extends BaseCollision {
	createCollider(isSensor: boolean = false) {
		//@ts-ignore
		let colliderDesc = ColliderDesc.heightfield(
			//@ts-ignore
			this.subdivisions, this.subdivisions, new Float32Array(this.heights), this.size
		);
		colliderDesc.setSensor(isSensor);
		colliderDesc.activeCollisionTypes = (isSensor) ? ActiveCollisionTypes.KINEMATIC_FIXED : ActiveCollisionTypes.DEFAULT;
		return colliderDesc;
	}
}