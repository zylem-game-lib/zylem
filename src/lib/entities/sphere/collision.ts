import { ActiveCollisionTypes, ColliderDesc } from "@dimforge/rapier3d-compat";
import { BaseCollision } from "~/lib/collision/_oldCollision";

export class SphereCollision extends BaseCollision {
	_collisionRadius: number = 1;

	createCollider(isSensor: boolean = false) {
		const radius = this._collisionRadius || 1;
		let colliderDesc = ColliderDesc.ball(radius);
		colliderDesc.setSensor(isSensor);
		colliderDesc.activeCollisionTypes = (isSensor) ? ActiveCollisionTypes.KINEMATIC_FIXED : ActiveCollisionTypes.DEFAULT;
		return colliderDesc;
	}
}
