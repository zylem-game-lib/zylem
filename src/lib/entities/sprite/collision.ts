import { ActiveCollisionTypes, ColliderDesc, RigidBodyDesc, RigidBodyType } from "@dimforge/rapier3d-compat";
import { Vector3 } from "three";
import { BaseCollision } from "~/lib/collision/_oldCollision";

export class SpriteCollision extends BaseCollision {
	collisionSize: Vector3 = new Vector3(1, 1, 1);
	size: Vector3 = new Vector3(1, 1, 1);

	createCollision({ isDynamicBody = true, isSensor = false }) {
		const gravityScale = (isSensor) ? 0.0 : 1.0;
		const type = isDynamicBody ? RigidBodyType.Dynamic : RigidBodyType.Fixed;
		this.bodyDescription = new RigidBodyDesc(type)
			.setTranslation(0, 0, 0)
			.lockRotations()
			.setGravityScale(gravityScale)
			.setCanSleep(false)
			.setCcdEnabled(true);
	}

	createCollider(isSensor: boolean = false) {
		const { x, y, z } = this.collisionSize ?? this.size;
		const size = new Vector3(x, y, z);
		const half = { x: size.x / 2, y: size.y / 2, z: size.z / 2 };
		let colliderDesc = ColliderDesc.cuboid(half.x, half.y, half.z);
		colliderDesc.setSensor(isSensor);
		colliderDesc.activeCollisionTypes = (isSensor) ? ActiveCollisionTypes.KINEMATIC_FIXED : ActiveCollisionTypes.DEFAULT;
		return colliderDesc;
	}
}