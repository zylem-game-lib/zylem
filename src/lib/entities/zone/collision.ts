import { ActiveCollisionTypes, ColliderDesc } from "@dimforge/rapier3d-compat";
import { Vector3 } from "three";
import { BaseCollision } from "~/lib/collision/_oldCollision";
import { SizeVector } from "~/lib/interfaces/utility";

export class ZoneCollision extends BaseCollision {
	_size: SizeVector = new Vector3(1, 1, 1);

	createCollider(_isSensor: boolean = true) {
		const size = this._size || new Vector3(1, 1, 1);
		const half = { x: size.x / 2, y: size.y / 2, z: size.z / 2 };
		let colliderDesc = ColliderDesc.cuboid(half.x, half.y, half.z);
		colliderDesc.setSensor(true);
		colliderDesc.activeCollisionTypes = ActiveCollisionTypes.KINEMATIC_FIXED;
		return colliderDesc;
	}
}