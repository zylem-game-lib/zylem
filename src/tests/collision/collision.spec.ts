import { describe, it, expect } from 'vitest';
import { BaseCollision } from '../../lib/collision/collision';

describe('create base collision', () => {
	it('should create collision component', () => {
		const result = new BaseCollision({});
		expect(result).toMatchSnapshot();
	});
});