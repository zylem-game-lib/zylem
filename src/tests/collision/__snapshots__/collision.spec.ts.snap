// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create base collision > should create collision component 1`] = `
BaseCollision {
  "bodyDescription": cA {
    "angularDamping": 0,
    "angularInertiaLocalFrame": oA {
      "w": 1,
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "angvel": iA {
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "canSleep": false,
    "ccdEnabled": true,
    "centerOfMass": iA {
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "dominanceGroup": 0,
    "enabled": true,
    "gravityScale": 1,
    "linearDamping": 0,
    "linvel": iA {
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "mass": 0,
    "massOnly": false,
    "principalAngularInertia": iA {
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "rotation": oA {
      "w": 1,
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "rotationsEnabledX": true,
    "rotationsEnabledY": true,
    "rotationsEnabledZ": true,
    "sleeping": false,
    "status": 0,
    "translation": {
      "x": 0,
      "y": 0,
      "z": 0,
    },
    "translationsEnabledX": true,
    "translationsEnabledY": true,
    "translationsEnabledZ": true,
  },
  "isSensor": false,
  "static": false,
}
`;
